bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogManagerForEntryLogPerLedger.java

```diff
@@ -0,0 +1,463 @@
+        Long ledgerId = expiredLedgerEntryLogMapEntry.getKey();
+        log.debug("LedgerId {} is not accessed for entrylogMapAccessExpiryTimeInSeconds"
+                + " period so it is being evicted from the cache map", ledgerId);
+        EntryLogAndLockTuple entryLogAndLockTuple = expiredLedgerEntryLogMapEntry.getValue();
+        Lock lock = entryLogAndLockTuple.ledgerLock;
+        BufferedLogChannel logChannel = entryLogAndLockTuple.entryLog;
+        lock.lock();
+        try {
+            replicaOfCurrentLogChannels.remove(logChannel.getLogId());
+            rotatedLogChannels.add(logChannel);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private LedgerDirsListener getLedgerDirsListener() {
+        return new LedgerDirsListener() {
+            @Override
+            public void diskFull(File disk) {
+                Set<BufferedLogChannel> copyOfCurrentLogs = getCopyOfCurrentLogs();
+                for (BufferedLogChannel currentLog : copyOfCurrentLogs) {
+                    if (disk.equals(currentLog.getLogFile().getParentFile())) {
+                        currentLog.setLedgerDirFull(true);
+                    }
+                }
+            }
+
+            @Override
+            public void diskWritable(File disk) {
+                Set<BufferedLogChannel> copyOfCurrentLogs = getCopyOfCurrentLogs();
+                for (BufferedLogChannel currentLog : copyOfCurrentLogs) {
+                    if (disk.equals(currentLog.getLogFile().getParentFile())) {
+                        currentLog.setLedgerDirFull(false);
+                    }
+                }
+            }
+        };
+    }
+
+    public void acquireLock(Long ledgerId) throws IOException {
+        try {
+            ledgerIdEntryLogMap.get(ledgerId).getLedgerLock().lock();
+        } catch (Exception e) {
+            log.error("Received unexpected exception while fetching lock to acquire", e);
+            throw new IOException("Received unexpected exception while fetching lock to acquire", e);
+        }
+    }
+
+    public void releaseLock(Long ledgerId) throws IOException {
+        try {
+            ledgerIdEntryLogMap.get(ledgerId).getLedgerLock().unlock();
+        } catch (Exception e) {
+            log.error("Received unexpected exception while fetching lock to release", e);
+            throw new IOException("Received unexpected exception while fetching lock to release", e);
+        }
+    }
+
+    /*
+     * sets the logChannel for the given ledgerId. It will add the new
+     * logchannel to replicaOfCurrentLogChannels, and the previous one will
+     * be removed from replicaOfCurrentLogChannels. Previous logChannel will
+     * be added to rotatedLogChannels in both the cases.
+     */
+    @Override
+    public void setCurrentLogForLedgerAndAddToRotate(long ledgerId, BufferedLogChannel logChannel) throws IOException {
+        acquireLock(ledgerId);
```

Comment:
what does the "lock" actually lock? does it prevent modifying the shared structure or the log channel itself? it is unclear to me. that says do we need to put the lock block for all the lines below?

Reply:
i followed simple rule, wherever in EntryLogManagerForSingleEntryLog, we are having code in 'synchronized' scope, i'm keeping the equivalent code of that in EntryLogManagerForEntryLogPerLedger, in guarded scope of ledgerId's lock.

so effectively ledgerId's lock is guarding everything related to this ledger - log channel and state information related to this ledger.